File: src/api/projectMerge.ts
// =============================================================================
// CORE-FFX - Forensic File Explorer
// Copyright (c) 2024-2026 CORE-FFX Project Contributors
// Licensed under MIT License - see LICENSE file for details
// =============================================================================

import { invoke } from "@tauri-apps/api/core";

// =============================================================================
// Types (match Rust merge module)
// =============================================================================

export interface ProjectMergeSummary {
  cffxPath: string;
  ffxdbPath: string;
  ffxdbExists: boolean;
  name: string;
  projectId: string;
  /** Owner/Examiner name from project owner_name field */
  ownerName: string | null;
  createdAt: string;
  savedAt: string;
  evidenceFileCount: number;
  hashCount: number;
  sessionCount: number;
  activityCount: number;
  bookmarkCount: number;
  noteCount: number;
  tabCount: number;
  reportCount: number;
  rootPath: string;
  /** Examiners/users found in .cffx and .ffxdb */
  examiners: MergeExaminerInfo[];
  /** Evidence collection summaries from .ffxdb */
  collections: MergeCollectionSummary[];
  /** COC item summaries from .ffxdb */
  cocItems: MergeCocSummary[];
  /** Form submission summaries from .ffxdb */
  formSubmissions: MergeFormSummary[];
  /** Evidence file summaries from .ffxdb */
  evidenceFiles: MergeEvidenceFileSummary[];
}

/** Examiner/user found in .cffx or .ffxdb */
export interface MergeExaminerInfo {
  name: string;
  displayName: string | null;
  /** Where this name was found: "cffx" or "ffxdb" */
  source: string;
  /** Role context: "project owner", "session user", "collecting officer", etc. */
  role: string;
}

/** Evidence collection summary from .ffxdb */
export interface MergeCollectionSummary {
  id: string;
  caseNumber: string;
  collectionDate: string;
  collectingOfficer: string;
  collectionLocation: string;
  status: string;
  itemCount: number;
}

/** COC item summary from .ffxdb */
export interface MergeCocSummary {
  id: string;
  cocNumber: string;
  caseNumber: string;
  evidenceId: string;
  description: string;
  submittedBy: string;
  receivedBy: string;
  status: string;
}

/** Form submission summary from .ffxdb */
export interface MergeFormSummary {
  id: string;
  templateId: string;
  caseNumber: string | null;
  status: string;
  createdAt: string;
  /** Collecting officer (from data_json for evidence_collection templates) */
  collectingOfficer: string | null;
  /** Collection location (from data_json for evidence_collection templates) */
  collectionLocation: string | null;
  /** Lead examiner (from data_json for IAR/activity templates) */
  leadExaminer: string | null;
}

/** Evidence file summary from .ffxdb */
export interface MergeEvidenceFileSummary {
  id: string;
  path: string;
  filename: string;
  containerType: string;
  totalSize: number;
}

/** Owner assignment for a source project during merge */
export interface MergeSourceAssignment {
  cffxPath: string;
  ownerName: string;
}

/** Provenance record for a merged source project */
export interface MergeSource {
  sourceProjectId: string;
  sourceProjectName: string;
  sourceCffxPath: string;
  ownerName: string;
  mergedAt: string;
  evidenceFileCount: number;
  sessionCount: number;
  activityCount: number;
  bookmarkCount: number;
  noteCount: number;
}

export interface MergeStats {
  projectsMerged: number;
  usersMerged: number;
  sessionsMerged: number;
  activityEntriesMerged: number;
  evidenceFilesMerged: number;
  hashesMerged: number;
  bookmarksMerged: number;
  notesMerged: number;
  tabsMerged: number;
  reportsMerged: number;
  tagsMerged: number;
  searchesMerged: number;
  ffxdbTablesMerged: number;
}

export interface MergeResult {
  success: boolean;
  cffxPath: string | null;
  ffxdbPath: string | null;
  error: string | null;
  stats: MergeStats | null;
  /** Provenance records for each source project */
  sources: MergeSource[] | null;
}

// =============================================================================
// API Functions
// =============================================================================

/** Analyze multiple .cffx files and return summaries for the merge wizard. */
export async function analyzeProjects(
  cffxPaths: string[],
): Promise<ProjectMergeSummary[]> {
  return invoke<ProjectMergeSummary[]>("project_merge_analyze", {
    cffxPaths,
  });
}

/** Execute a full project merge. */
export async function executeMerge(
  cffxPaths: string[],
  outputPath: string,
  mergedName: string,
  newRoot?: string,
  ownerAssignments?: MergeSourceAssignment[],
  excludeCollectionIds?: string[],
): Promise<MergeResult> {
  return invoke<MergeResult>("project_merge_execute", {
    cffxPaths,
    outputPath,
    mergedName,
    newRoot: newRoot ?? null,
    ownerAssignments: ownerAssignments ?? null,
    excludeCollectionIds: excludeCollectionIds ?? null,
  });
}

File: src-tauri/src/commands/project_merge.rs
// =============================================================================
// CORE-FFX - Forensic File Explorer
// Copyright (c) 2024-2026 CORE-FFX Project Contributors
// Licensed under MIT License - see LICENSE file for details
// =============================================================================

//! Tauri commands for project merge and relocate operations.

use crate::project::merge;

/// Analyze multiple .cffx files and return summaries for the merge wizard.
#[tauri::command]
pub async fn project_merge_analyze(cffx_paths: Vec<String>) -> Vec<merge::ProjectMergeSummary> {
    tauri::async_runtime::spawn_blocking(move || merge::analyze_projects(&cffx_paths))
        .await
        .unwrap_or_default()
}

/// Execute a full project merge pipeline.
///
/// - `cffx_paths`: list of source .cffx file paths to merge
/// - `output_path`: where to save the merged .cffx
/// - `merged_name`: name for the merged project
/// - `new_root`: optional new root directory (for relocation)
/// - `owner_assignments`: optional owner/examiner assignments for each source project
/// - `exclusions`: optional exclusion configuration for selective merging
///   (skip entire categories, individual evidence files, COC items, collections, or form submissions)
#[tauri::command]
pub async fn project_merge_execute(
    cffx_paths: Vec<String>,
    output_path: String,
    merged_name: String,
    new_root: Option<String>,
    owner_assignments: Option<Vec<merge::MergeSourceAssignment>>,
    exclusions: Option<merge::MergeExclusions>,
) -> merge::MergeResult {
    tauri::async_runtime::spawn_blocking(move || {
        merge::execute_merge(
            &cffx_paths,
            &output_path,
            &merged_name,
            new_root.as_deref(),
            owner_assignments.as_deref(),
            exclusions.as_ref(),
        )
    })
    .await
    .unwrap_or_else(|e| merge::MergeResult {
        success: false,
        cffx_path: None,
        ffxdb_path: None,
        error: Some(format!("Task failed: {}", e)),
        stats: None,
        sources: None,
    })
}
